Adding new function pv_angle_corr() to panels.py, to apply correction…

benjamin-saliwanchik opened the pull request:
… for PV departure from cosine law.

benjamin-saliwanchik committed:
Adding new function pv_angle_corr() to panels.py, to apply correction for PV departure from cosine law.

slosar commented on hardware/panels.py:
```diff
@@ -57,7 +72,7 @@ def exposure(self):
         eff = 0.3 # default, if the temperature curve is not set for the sun
         if self.sun.regolith_temperature is not None: eff = self.pvEfficiency(self.sun.regolith_temperature)
-        power =  Panel.solarConstant*eff*np.select(self.condition_list, self.choice_list) * self.efficiency_mult
+        power =  Panel.solarConstant*eff*np.select(self.condition_list, self.choice_list)*self.pv_angle_corr * self.efficiency_mult
```
Does this actually work? You're  calling it as a variable, not as a function? Or am I missing something?

benjamin-saliwanchik replied:
Yep. In line 37 the variable version is defined from the function version with the current sun coordinates:         self.pv_angle_corr = self.pv_angle_corr(sun.xyz). That was the way Maxim had the dot product and normal vector rotation defined, I think for simplicity of handling, so I followed that convention.

slosar replied:
Ok, but this is kinda super dangerous, because if nothing else interpreter would replace the function with a variable. Can you just say instead something like `self.pv_angle_corr_calculated = self.pv_angle_corr(sun.xyz)`?

benjamin-saliwanchik replied:
Good point. Variable name changed to dot_sun_corr since it's a correction to the sun-PV dot product, which is called dot_sun.

slosar commented on hardware/panels.py:
```diff
@@ -43,6 +44,20 @@ def dot(self, sun_xyz):
+
+    ### ---
+    def pv_angle_corr(self, sun_xyz):
+        #Correction factor for measured divergence of PV power as a function of solar angle from expected cosine/dot-product dependence 
```
Add a note to self here that will allow you to come back to this in 10 years. I.e. something like "See Ben Sawlianchik's private blablah.ipynb for how to the fitting function was determined."

benjamin-saliwanchik replied:
Will add in next commit. I will also add the "apply_cos_correction" to the devices.yml file and the __init__ function for the panel class.

slosar replied:
Yes, keep the PR open, I should see your update commits.

slosar reviewed (changes requested):
In general looks good, alhough I'm not sure it works. 

Please make an "apply_cos_correction" option to the constructor and then make an option in the yaml file to enable or disable this.

benjamin-saliwanchik committed:
Added apply_cosine_correction option to panel constructor.

Added cosine correction option, set in devices.yml. Renamed correction variable to dot_sun_corr to avoid overwriting func name. Updated doc strings.

benjamin-saliwanchik committed:
Merge branch 'main' into 39-revise-solar-panels-with-departures-from-cosine-law

benjamin-saliwanchik committed:
Merging main into 39-revise-solar... branch

benjamin-saliwanchik committed:
Update reference_data in sim_test.py to include pv_angle_corr().

slosar reviewed (approved):
Looks good.
